### PR TITLE
Implement "allow keys" authorization handler

### PR DESCRIPTION
**Pull request by ltseeley**
Implements the `AllowKeysAuthorizationHandler` for specifying admin keys in a file

### Testing

1. Startup splinterd with experimental features:

```bash
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- database migrate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- cert generate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features experimental --  --rest-api-endpoint http://localhost:8080 --tls-insecure -vv
```

2. Generate a key pair for yourself if you don't already have one:

```bash
$ cargo run --manifest-path cli/Cargo.toml -- keygen
```

3. Attempt to list the node's circuits and verify the request is unauthorized:

```bash
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- circuit list -U http://localhost:8080
ERROR: Subcommand encountered an error: Failed to list circuits: Client is not authorized
```

4. Add your public key to the allowed keys list:

```bash
$ cat ~/.splinter/keys/<YOUR-USERNAME>.pub > /tmp/splinter/data/allow_keys
```

5. Verify that you can now list the node's circuits (even though the list is empty):

```bash
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- circuit list -U http://localhost:8080
ID NAME MANAGEMENT MEMBERS
```

### CODE REVIEW

File: libsplinter/src/rest_api/auth/allow_keys.rs [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+/// the backing file, the handler will return [`AuthorizationHandlerResult::Allow`]; if not, it will
+/// return [`AuthorizationHandlerResult::Continue`]. The `permission_id` argument for
+/// [`AuthorizationHandler::has_permission`] is ignored because this authorization handler provides
+/// admin privileges (all permissions).
+///
+/// The authorization handler's backing file must be a list of valid Cylinder [`PublicKey`]s,
+/// separated by newlines.
+///
+/// The list of keys in the file are cached in-memory by the authorization handler; this means that
+/// the handler will continue to work even if the backing file becomes unavailable. Each time the
```
Comment (ltseeley) [inline libsplinter/src/rest_api/auth/allow_keys.rs:43]: Interested in opinions on whether or not the handler should tolerate the file being removed

Comment (vaporos) [reply libsplinter/src/rest_api/auth/allow_keys.rs:43]: In this auth case, it probably would make sense to treat it the same as an empty file.

File: libsplinter/src/rest_api/auth/allow_keys.rs [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+                "failed to open allow keys file".into(),
+            )
+        })?;
+        let keys = BufReader::new(file)
+            .lines()
+            .filter_map(|res| match res {
+                Ok(line) => match PublicKey::new_from_hex(&line) {
+                    Ok(key) => Some(key),
+                    Err(err) => {
+                        error!("Invalid key in allow keys file: {}", err);
```
Comment (peterschwarz) [inline libsplinter/src/rest_api/auth/allow_keys.rs:176]: If you were to enumerate the iterator, you could log which line the bad key was on (the err probably won't give any details about which key it was, just what was wrong with the key)

File: libsplinter/src/rest_api/auth/allow_keys.rs (hunk 1 above)
Comment (ltseeley) [reply libsplinter/src/rest_api/auth/allow_keys.rs]: Yeah that makes more sense; updated.

File: libsplinter/src/rest_api/auth/allow_keys.rs (hunk 2 above)
Comment (ltseeley) [reply libsplinter/src/rest_api/auth/allow_keys.rs]: Good idea; updated.

File: libsplinter/src/rest_api/auth/allow_keys.rs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            )
+        })?;
+        let keys = BufReader::new(file)
+            .lines()
+            .enumerate()
+            .filter_map(|(idx, res)| match res {
+                Ok(line) => match PublicKey::new_from_hex(&line) {
+                    Ok(key) => Some(key),
+                    Err(err) => {
+                        error!("Invalid key on line {} of allow keys file: {}", idx, err);
```
Comment (peterschwarz) [inline libsplinter/src/rest_api/auth/allow_keys.rs:177]: Lines are 1-indexed, enumerate is 0-indexed

Comment (ltseeley) [reply libsplinter/src/rest_api/auth/allow_keys.rs]: Didn't realize that; fixed.

**Review by peterschwarz — DISMISSED** [review]

**Review by vaporos — DISMISSED** [review]

**Review by peterschwarz — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**